### ISSUE TITLE
[FR] git-daemon-export-ok

### ISSUE THREAD

**Issue by ghost** [gogits/gogs#2940]
Would it be possible to add an option to ensure presence of `git-daemon-export-ok` file for public repositories and its absence for private ones? The presence of this file guides the `git-daemon` used to provide `git://` protocol as to which repositories to actually expose.

AFAICT, using `git-daemon` to expose repos read-only in parallel with gogs works, but there is no distinction for private / public ones, so it exposes either none of them (cause the file is not present), or all of them (if run with `--export-all`).

**Comment by strk** [comment]
See also #2950 for another problem related to Gogs making no distinction between private / public repos.

**Comment by strk** [comment]
Looking at the code the support for adding/removing `git-export-ok` should be added to https://github.com/gogits/git-module, I suggest you file an issue there @herby

**Comment by herby** [comment]
@strk Why there? It depends on the private/public setting of the gogs itself, it is not very much of a git operation... ping @Unknwon.

**Comment by strk** [comment]
Because that's the module used to interact with the git repo.
So that module should expose a function to add/remove the git-daemon-export-ok
and Gogs itself should be using that function.

Ideally the module would also expose functions to set the repository ownership and permissions, to deal with #2950

**Comment by herby** [comment]
I see.

Dňa 16. apríla 2016 19:10:37 CEST používateľ Sandro Santilli <notifications@github.com> napísal:

**Comment by Unknwon** [comment]
Hmm... any reference for `git-daemon-export-ok`, did not know about this feature of git before.

**Comment by herby** [comment]
It's feature of `git daemon`, https://git-scm.com/book/en/v2/Git-on-the-Server-Git-Daemon.

无闻 wrote:

**Comment by Unknwon** [comment]
Thanks for the link. I think we need to let Gogs support `git://` protocol first?

**Comment by herby** [comment]
I don't know what do you mean by that. I asked you via comment on faq if I can run git-daemon in parallel and you said it was ok. This would just tell it which o es to expose and which not.
If you mean to support git:// protocol directly by gogs code, that would be completely different issue.

**Comment by strk** [comment]
I agree there's no need for Gogs to support git:// protocol itself
(although it could be useful to _allow_ adding an indication of git://
access method when available by external means).

On the other hand, having access mediated by different servers is
always a risk, unless all servers carefully agree on what kind
of access level to grant to the resources (I'm thinking about the
"require sign in" setting of Gogs for public repos, for example).

The usage of file-based configurations (like git-export-ok) is a good
way to encode permission levels at the filesystem level. Maybe Gogs
could use a similar mechanism for the "require-sign-in" config.

**Comment by Unknwon** [comment]
> if I can run git-daemon in parallel and you said it was ok

Yes, if you run parallel, that's 100% OK, but now you ask Gogs to control private/public, and that is **completely different issue**.

**Comment by herby** [comment]
??? Confused. Don't see "two steps".

But yes, I am asking about the option to add/remove `git-daamon-export-ok` file depending on whether repo is set private or public in gogs.

Is there some misunderstanding (you told "I think we need to let Gogs support git:// protocol first?" I don't know what you mean by that. No "support first" is needed, though I don't know what do you mean by that)? Just to add/remove the file that help git-daemon control the exposing, if used.

**Comment by Unknwon** [comment]
@herby Gogs does no support `git://` provided by `git-daemon`, how should Gogs care/manage settings for it?

**Comment by herby** [comment]
I don't know what's the communication problem here, but there must be some.

What I ask in this issue is the creation of one empty file for a public(ized) repo and deleting that file for private(ized) repo.

The use case is, of course, to allow to run git daemon in parallel (read-only) which works by your own words.

So I only ask for that one flag file management.

Is the problem that you must "declare that gogs supports git:// protocol" to implement the issue? I think not. Yoy can just add it for convenience of those who do use the protocol, but still not saying "gogs supports ...".

Or is there another problem? I am a bit desperate here to spot one, but I see you see it there, when I only ask for the silly adding/removing one specific file flagging the public state of repo, but you are always telling about supporting git protocol (you do not need to, git daemon does all the work).

Dňa 26. apríla 2016 5:56:46 CEST používateľ "无闻" <notifications@github.com> napísal:

**Comment by Unknwon** [comment]
cc @bkcsoft : can you work on this one? I think it needs some changes in `git-module` as well.

**Comment by bkcsoft** [comment]
Yeah sure, should be fairly simple :) Don't think I have to involve
git-module at all TBH

**Comment by bkcsoft** [comment]
Done #3046
(@Unknwon and no need for fiddling in `git-module` :trollface:)

**Comment by herby** [comment]
BTW if this https://github.com/bower/bower/issues/1841 ends up positively, maybe this issue could be put on hold, if needed (but I still did not get an answer). I only needed `git://` protocol was only protocol allowed in bower.

**Comment by Unknwon** [comment]
Patch has pushed to fix this issue, please test on `develop` branch or https://try.gogs.io.
